### PR TITLE
LocalisedTransaction && LocalisedTransactionReceipt

### PR DESCRIPTION
**Pull request by debris**
An attempt to move unnecessary logic from WebThreeStubServerBase to Client.

### CODE REVIEW

File: libethereum/BlockChain.h [hunk 1]
```diff
@@ -142,8 +142,11 @@ class BlockChain
 	BlockReceipts receipts(h256 const& _hash) const { return queryExtras<BlockReceipts, ExtraReceipts>(_hash, m_receipts, x_receipts, NullBlockReceipts); }
 	BlockReceipts receipts() const { return receipts(currentHash()); }
 
+	/// Get the transaction by block hash and index;
+	TransactionReceipt transactionReceipt(h256 const& _blockHash, unsigned _i) const {return receipts(_blockHash).receipts[_i]; }
```
Comment (gavofyork) [inline libethereum/BlockChain.h]: spacing.

File: libethereum/BlockChain.h [hunk 2]
```diff
@@ -142,8 +142,11 @@ class BlockChain
...
+
 	/// Get the transaction receipt by transaction hash. Thread-safe.
-	TransactionReceipt transactionReceipt(h256 const& _transactionHash) const {TransactionAddress ta = queryExtras<TransactionAddress, ExtraTransactionAddress>(_transactionHash, m_transactionAddresses, x_transactionAddresses, NullTransactionAddress); if (!ta) return bytesConstRef(); return receipts(ta.blockHash).receipts[ta.index]; }
+	TransactionReceipt transactionReceipt(h256 const& _transactionHash) const {TransactionAddress ta = queryExtras<TransactionAddress, ExtraTransactionAddress>(_transactionHash, m_transactionAddresses, x_transactionAddresses, NullTransactionAddress); if (!ta) return bytesConstRef(); return transactionReceipt(ta.blockHash, ta.index); }
```
Comment (gavofyork) [inline libethereum/BlockChain.h]: spacing

File: libethereum/ClientBase.cpp [hunk 3]
```diff
@@ -317,6 +317,12 @@ Transaction ClientBase::transaction(h256 _transactionHash) const
 	return Transaction(bc().transaction(_transactionHash), CheckTransaction::Cheap);
 }
 
+LocalisedTransaction ClientBase::localisedTransaction(h256 _transactionHash) const
```
Comment (gavofyork) [inline libethereum/ClientBase.cpp]: const&

File: libethereum/ClientBase.cpp [hunk 4]
```diff
@@ -327,11 +333,31 @@ Transaction ClientBase::transaction(h256 _blockHash, unsigned _i) const
 		return Transaction();
 }
 
+LocalisedTransaction ClientBase::localisedTransaction(h256 _blockHash, unsigned _i) const
```
Comment (gavofyork) [inline libethereum/ClientBase.cpp]: const&

File: libethereum/ClientBase.h [hunk 5]
```diff
@@ -118,8 +118,11 @@ class ClientBase: public Interface
 	virtual BlockInfo blockInfo(h256 _hash) const override;
 	virtual BlockDetails blockDetails(h256 _hash) const override;
 	virtual Transaction transaction(h256 _transactionHash) const override;
+	virtual LocalisedTransaction localisedTransaction(h256 _transactionHash) const override;
```
Comment (gavofyork) [inline libethereum/ClientBase.h]: const&

File: libethereum/ClientBase.h [hunk 6]
```diff
@@ -118,8 +118,11 @@ class ClientBase: public Interface
...
 	virtual Transaction transaction(h256 _blockHash, unsigned _i) const override;
+	virtual LocalisedTransaction localisedTransaction(h256 _blockHash, unsigned _i) const override;
```
Comment (gavofyork) [inline libethereum/ClientBase.h]: const&

File: libethereum/ClientBase.h [hunk 7]
```diff
@@ -118,8 +118,11 @@ class ClientBase: public Interface
...
 	virtual TransactionReceipt transactionReceipt(h256 const& _transactionHash) const override;
+	virtual LocalisedTransactionReceipt localisedTransactionReceipt(h256 const& _transactionHash) const override;
```
Comment (gavofyork) [inline libethereum/ClientBase.h]: const&

**Comment by gavofyork** [comment]
aside from style stuff, would be nice to remove `pair<h256, unsigned>` as the return from `transactionLocation` and either have a struct or (better) use a modified version of `LocalisedTransaction`, where the block number component is allowed to be undefined.

**Comment by debris** [comment]
there is already a struct `[TransactionAddress](https://github.com/ethereum/cpp-ethereum/blob/develop/libethereum/BlockDetails.h#L99)` which contains exactly the same info as `pair<h256, unsigned>`. Can we use it? I will create separate pr for this.

**Comment by gavofyork** [comment]
mac build fail?

**Comment by debris** [comment]
it's unrelated.

**Comment by gavofyork** [comment]
rebase?

**Comment by debris** [comment]
It's already merged as a part of #2504

**Pull request closed without merging**